File: analyzer/src/utils/workflow_state_manager.py
"""
Workflow state manager for bidirectional workflows.

Provides atomic read/write operations for workflow-state.json,
which tracks the execution state of analyze, audit, plan, and improve commands.
"""

import json
from pathlib import Path

from ..models.workflow_state import CommandState, WorkflowState
from ..models.workflow_state_migrations import (
    CURRENT_WORKFLOW_STATE_VERSION,
    apply_migrations,
)
from .state_manager import StateManager


class WorkflowStateManager:
    """
    Manages workflow state persistence for bidirectional workflows.

    Provides atomic read/write operations using temp file + rename pattern.
    """

    @staticmethod
    def _get_workflow_state_file(base_path: Path | None = None) -> Path:
        """Get the path to the workflow state file.

        Args:
            base_path: Base directory for .docimp state files (default: cwd)
        """
        return StateManager.get_state_dir(base_path) / "workflow-state.json"

    @staticmethod
    def save_workflow_state(state: WorkflowState) -> None:
        """
        Save workflow state to disk atomically (temp file + rename pattern).

        Args:
            state: The workflow state to save

        Raises:
            IOError: If file write fails
        """
        file_path = WorkflowStateManager._get_workflow_state_file()
        temp_path = file_path.with_suffix(".json.tmp")

        # Ensure state directory exists
        StateManager.ensure_state_dir()

        # Serialize to JSON
        data = state.to_dict()

        # Write to temp file first
        with temp_path.open("w", encoding="utf-8") as f:
            json.dump(data, f, indent=2)

        # Atomic rename
        temp_path.replace(file_path)

    @staticmethod
    def load_workflow_state(base_path: Path | None = None) -> WorkflowState:
        """
        Load workflow state from disk with schema validation and migration support.
        Returns empty state if file doesn't exist.

        Args:
            base_path: Base directory for .docimp state files (default: cwd)

        Returns:
            WorkflowState: The loaded or empty workflow state

        Raises:
            ValueError: If JSON is malformed or schema is invalid
        """
        file_path = WorkflowStateManager._get_workflow_state_file(base_path)

        if not file_path.exists():
            return WorkflowState.create_empty()

        try:
            with file_path.open(encoding="utf-8") as f:
                data = json.load(f)

            # Apply migrations if needed (handles legacy files and version upgrades)
            migrated = apply_migrations(data, CURRENT_WORKFLOW_STATE_VERSION)

            return WorkflowState.from_dict(migrated)
        except (json.JSONDecodeError, KeyError) as e:
            raise ValueError(f"Failed to load workflow state: {e}") from e

    @staticmethod
    def update_command_state(
        command: str,
        command_state: CommandState,
        *,
        history_enabled: bool = False,
        history_max_snapshots: int = 50,
        history_max_age_days: int = 30,
    ) -> None:
        """
        Update the state for a specific command.

        Args:
            command: The command name ('analyze', 'audit', 'plan', 'improve')
            command_state: The command state to save
            history_enabled: Enable workflow history tracking (default: False)
            history_max_snapshots: Maximum number of snapshots to keep (default: 50)
            history_max_age_days: Maximum age in days to keep snapshots (default: 30)

        Raises:
            ValueError: If command name is invalid
        """
        valid_commands = {"analyze", "audit", "plan", "improve"}
        if command not in valid_commands:
            raise ValueError(
                f"Invalid command: {command}. Must be one of {valid_commands}"
            )

        state = WorkflowStateManager.load_workflow_state()

        # Update the specific command state
        if command == "analyze":
            state.last_analyze = command_state
        elif command == "audit":
            state.last_audit = command_state
        elif command == "plan":
            state.last_plan = command_state
        elif command == "improve":
            state.last_improve = command_state

        WorkflowStateManager.save_workflow_state(state)

        # Save history snapshot if enabled
        if history_enabled:
            WorkflowStateManager.save_history_snapshot(state)

            # Rotate old snapshots using hybrid strategy
            WorkflowStateManager.rotate_history(
                max_snapshots=history_max_snapshots,
                max_age_days=history_max_age_days,
            )

    @staticmethod
    def get_command_state(command: str) -> CommandState | None:
        """
        Get the state for a specific command.

        Args:
            command: The command name ('analyze', 'audit', 'plan', 'improve')

        Returns:
            CommandState or None if command hasn't been run

        Raises:
            ValueError: If command name is invalid
        """
        valid_commands = {"analyze", "audit", "plan", "improve"}
        if command not in valid_commands:
            raise ValueError(
                f"Invalid command: {command}. Must be one of {valid_commands}"
            )

        state = WorkflowStateManager.load_workflow_state()

        command_map = {
            "analyze": state.last_analyze,
            "audit": state.last_audit,
            "plan": state.last_plan,
            "improve": state.last_improve,
        }

        return command_map.get(command)

    @staticmethod
    def clear_workflow_state() -> None:
        """Delete the workflow state file if it exists."""
        file_path = WorkflowStateManager._get_workflow_state_file()

        if file_path.exists():
            file_path.unlink()

    @staticmethod
    def exists() -> bool:
        """Check if workflow state file exists."""
        return WorkflowStateManager._get_workflow_state_file().exists()

    @staticmethod
    def save_history_snapshot(state: WorkflowState) -> Path:
        """
        Save a timestamped snapshot of workflow state to history directory.
        Uses atomic write pattern (temp + rename) for safety.

        Args:
            state: The workflow state to snapshot

        Returns:
            Path to the created snapshot file

        Raises:
            IOError: If file write fails
        """
        from datetime import UTC, datetime

        # Generate cross-platform safe timestamp (replace : and . with -)
        timestamp = (
            datetime.now(tz=UTC)
            .isoformat()
            .replace(":", "-")
            .replace(".", "-")
        )
        filename = f"workflow-state-{timestamp}.json"
        filepath = StateManager.get_history_dir() / filename
        temp_path = filepath.with_suffix(".json.tmp")

        # Ensure history directory exists
        StateManager.ensure_state_dir()

        # Serialize to JSON
        data = state.to_dict()

        # Write to temp file first
        with temp_path.open("w", encoding="utf-8") as f:
            json.dump(data, f, indent=2)

        # Atomic rename
        temp_path.replace(filepath)

        return filepath

    @staticmethod
    def list_history_snapshots() -> list[Path]:
        """
        List all workflow state history snapshots, sorted newest first.

        Returns:
            List of snapshot file paths, sorted by timestamp (newest first)
        """
        history_dir = StateManager.get_history_dir()

        if not history_dir.exists():
            return []

        # Filter for workflow state snapshots only
        snapshots = [
            f
            for f in history_dir.iterdir()
            if f.is_file()
            and f.name.startswith("workflow-state-")
            and f.name.endswith(".json")
        ]

        # Sort by filename (ISO 8601 timestamps are lexicographically sortable)
        # Reverse to get newest first
        return sorted(snapshots, reverse=True)

    @staticmethod
    def rotate_history(max_snapshots: int = 50, max_age_days: int = 30) -> None:
        """
        Rotate workflow history using hybrid strategy:
        - Keep last N snapshots (count limit)
        - Keep snapshots from last M days (time limit)
        - Delete snapshots that violate BOTH limits

        Args:
            max_snapshots: Maximum number of snapshots to keep (default: 50)
            max_age_days: Maximum age in days to keep snapshots (default: 30)
        """
        import time

        snapshots = WorkflowStateManager.list_history_snapshots()

        if not snapshots:
            return  # Nothing to rotate

        # Calculate age threshold (Unix timestamp in seconds)
        now = time.time()
        age_threshold = now - (max_age_days * 24 * 60 * 60)

        to_delete: list[Path] = []

        for i, snapshot in enumerate(snapshots):
            # Get file modification time
            file_age = snapshot.stat().st_mtime

            # Hybrid logic: Delete if BOTH conditions are violated
            violates_count_limit = i >= max_snapshots  # Index-based (0-indexed)
            violates_time_limit = file_age < age_threshold

            if violates_count_limit or violates_time_limit:
                to_delete.append(snapshot)

        # Delete snapshots that violate limits
        for snapshot in to_delete:
            snapshot.unlink()

File: cli/src/utils/workflow-state-manager.ts
import * as fs from 'node:fs/promises';
import path from 'node:path';
import type { IWorkflowHistoryConfig } from '../config/i-config.js';
import {
  applyMigrations,
  CURRENT_WORKFLOW_STATE_VERSION,
} from '../types/workflow-state-migrations.js';
import {
  WorkflowState,
  WorkflowStateSchema,
  CommandState,
  createEmptyWorkflowState,
} from '../types/workflow-state.js';
import { StateManager } from './state-manager.js';

/**
 * Manages workflow state persistence for bidirectional workflows.
 *
 * Provides atomic read/write operations for workflow-state.json,
 * which tracks the execution state of analyze, audit, plan, and improve commands.
 */
export class WorkflowStateManager {
  /**
   * Get the path to the workflow state file
   *
   * @returns Path to workflow-state.json
   */
  private static getWorkflowStateFile(): string {
    return path.join(StateManager.getStateDir(), 'workflow-state.json');
  }

  /**
   * Save workflow state to disk atomically (temp file + rename pattern)
   */
  static async saveWorkflowState(state: WorkflowState): Promise<void> {
    const filePath = this.getWorkflowStateFile();
    const temporaryPath = `${filePath}.tmp`;

    // Ensure state directory exists
    StateManager.ensureStateDir();

    // Validate state against schema
    const validated = WorkflowStateSchema.parse(state);

    // Write to temp file first
    await fs.writeFile(
      temporaryPath,
      JSON.stringify(validated, null, 2),
      'utf8'
    );

    // Atomic rename
    await fs.rename(temporaryPath, filePath);
  }

  /**
   * Load workflow state from disk with schema validation and migration support
   * Returns empty state if file doesn't exist
   *
   * @returns Loaded workflow state or empty state if file doesn't exist
   */
  static async loadWorkflowState(): Promise<WorkflowState> {
    const filePath = this.getWorkflowStateFile();

    try {
      const content = await fs.readFile(filePath, 'utf8');
      const data = JSON.parse(content);

      // Apply migrations if needed (handles legacy files and version upgrades)
      const migrated = applyMigrations(data, CURRENT_WORKFLOW_STATE_VERSION);

      // Validate against schema (Zod will provide detailed validation errors)
      return WorkflowStateSchema.parse(migrated);
    } catch (error: unknown) {
      if (
        error instanceof Error &&
        'code' in error &&
        error.code === 'ENOENT'
      ) {
        // File doesn't exist - return empty state
        return createEmptyWorkflowState();
      }
      const message = error instanceof Error ? error.message : String(error);
      throw new Error(`Failed to load workflow state: ${message}`);
    }
  }

  /**
   * Update the state for a specific command
   *
   * @param command - Command to update state for
   * @param commandState - New state for the command
   * @param historyConfig - Optional workflow history configuration for snapshot tracking
   */
  static async updateCommandState(
    command: 'analyze' | 'audit' | 'plan' | 'improve',
    commandState: CommandState,
    historyConfig?: IWorkflowHistoryConfig
  ): Promise<void> {
    const state = await this.loadWorkflowState();

    // Update the specific command state
    switch (command) {
      case 'analyze': {
        state.last_analyze = commandState;
        break;
      }
      case 'audit': {
        state.last_audit = commandState;
        break;
      }
      case 'plan': {
        state.last_plan = commandState;
        break;
      }
      case 'improve': {
        state.last_improve = commandState;
        break;
      }
    }

    await this.saveWorkflowState(state);

    // Save history snapshot if enabled
    if (historyConfig?.enabled) {
      await this.saveHistorySnapshot(state);

      // Rotate old snapshots using hybrid strategy
      const maxSnapshots = historyConfig.maxSnapshots ?? 50;
      const maxAgeDays = historyConfig.maxAgeDays ?? 30;
      await this.rotateHistory(maxSnapshots, maxAgeDays);
    }
  }

  /**
   * Get the state for a specific command
   *
   * @returns Command state or null if not run yet
   */
  static async getCommandState(
    command: 'analyze' | 'audit' | 'plan' | 'improve'
  ): Promise<CommandState | null> {
    const state = await this.loadWorkflowState();

    switch (command) {
      case 'analyze': {
        return state.last_analyze;
      }
      case 'audit': {
        return state.last_audit;
      }
      case 'plan': {
        return state.last_plan;
      }
      case 'improve': {
        return state.last_improve;
      }
    }
  }

  /**
   * Delete the workflow state file
   */
  static async clearWorkflowState(): Promise<void> {
    const filePath = this.getWorkflowStateFile();

    try {
      await fs.unlink(filePath);
    } catch (error: unknown) {
      if (
        error instanceof Error &&
        'code' in error &&
        error.code !== 'ENOENT'
      ) {
        throw error;
      }
      // File doesn't exist - nothing to do
    }
  }

  /**
   * Check if workflow state file exists
   *
   * @returns True if workflow state file exists
   */
  static async exists(): Promise<boolean> {
    const filePath = this.getWorkflowStateFile();
    try {
      await fs.access(filePath);
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Save a timestamped snapshot of workflow state to history directory
   * Uses atomic write pattern (temp + rename) for safety
   *
   * @param state - Workflow state to snapshot
   * @returns Path to the created snapshot file
   */
  static async saveHistorySnapshot(state: WorkflowState): Promise<string> {
    // Generate cross-platform safe timestamp (replace : and . with -)
    const timestamp = new Date().toISOString().replaceAll(/[:.]/g, '-');
    const filename = `workflow-state-${timestamp}.json`;
    const filepath = path.join(StateManager.getHistoryDir(), filename);
    const temporaryPath = `${filepath}.tmp`;

    // Ensure history directory exists
    StateManager.ensureStateDir();

    // Validate state against schema
    const validated = WorkflowStateSchema.parse(state);

    // Write to temp file first
    await fs.writeFile(
      temporaryPath,
      JSON.stringify(validated, null, 2),
      'utf8'
    );

    // Atomic rename
    await fs.rename(temporaryPath, filepath);

    return filepath;
  }

  /**
   * List all workflow state history snapshots, sorted newest first
   *
   * @returns Array of snapshot file paths, sorted by timestamp (newest first)
   */
  static async listHistorySnapshots(): Promise<string[]> {
    const historyDirectory = StateManager.getHistoryDir();

    try {
      const files = await fs.readdir(historyDirectory);

      // Filter for workflow state snapshots only
      const snapshots = files
        .filter((f) => f.startsWith('workflow-state-') && f.endsWith('.json'))
        .map((f) => path.join(historyDirectory, f))
        .toSorted()
        .toReversed(); // ISO 8601 timestamps are lexicographically sortable

      return snapshots;
    } catch (error: unknown) {
      if (
        error instanceof Error &&
        'code' in error &&
        error.code === 'ENOENT'
      ) {
        // History directory doesn't exist yet - return empty array
        return [];
      }
      throw error;
    }
  }

  /**
   * Rotate workflow history using hybrid strategy:
   * - Keep last N snapshots (count limit)
   * - Keep snapshots from last M days (time limit)
   * - Delete snapshots that violate BOTH limits
   *
   * @param maxSnapshots - Maximum number of snapshots to keep (default: 50)
   * @param maxAgeDays - Maximum age in days to keep snapshots (default: 30)
   */
  static async rotateHistory(
    maxSnapshots: number = 50,
    maxAgeDays: number = 30
  ): Promise<void> {
    const snapshots = await this.listHistorySnapshots();

    if (snapshots.length === 0) {
      return; // Nothing to rotate
    }

    // Calculate age threshold (Unix timestamp in milliseconds)
    const now = Date.now();
    const ageThresholdMs = maxAgeDays * 24 * 60 * 60 * 1000;
    const ageThreshold = now - ageThresholdMs;

    const toDelete: string[] = [];

    for (const [i, snapshot] of snapshots.entries()) {
      // Get file modification time
      const stats = await fs.stat(snapshot);
      const fileAge = stats.mtimeMs;

      // Hybrid logic: Delete if BOTH conditions are violated
      const violatesCountLimit = i >= maxSnapshots; // Index-based (0-indexed)
      const violatesTimeLimit = fileAge < ageThreshold;

      if (violatesCountLimit || violatesTimeLimit) {
        toDelete.push(snapshot);
      }
    }

    // Delete snapshots that violate limits
    await Promise.all(toDelete.map((snapshot) => fs.unlink(snapshot)));
  }
}
